pkg/controller/knativeserving/openshift/openshift.go

```diff
@@ -167,34 +178,146 @@ func checkVersion(instance *servingv1alpha1.KnativeServing) error {
 	return nil
 }
 
-// Check for service mesh memberRoll for maistra
-func verifyServiceMeshMemberRoll(instance *servingv1alpha1.KnativeServing) error {
-	smmr := &maistrav1.ServiceMeshMemberRoll{}
-	checkError := func(err error, instance *servingv1alpha1.KnativeServing) error {
-		msg := fmt.Sprintf("Istio not detected; please install ServiceMesh")
-		instance.Status.MarkDependencyMissing(msg)
-		log.Error(err, msg)
+func ingressNamespace(servingNamespace string) string {
+	return servingNamespace + "-ingress"
+}
+
+func createIngressNamespace(servingNamespace string) error {
+	ns := &v1.Namespace{}
+	if err := api.Get(context.TODO(), client.ObjectKey{Name: ingressNamespace(servingNamespace)}, ns); err != nil {
+		if apierrors.IsNotFound(err) {
+			ns.Name = ingressNamespace(servingNamespace)
+			if err = api.Create(context.TODO(), ns); err != nil {
+				return err
+			}
+			return nil
+		}
 		return err
 	}
-	if err := api.Get(context.TODO(), client.ObjectKey{Namespace: "istio-system", Name: "default"}, smmr); err != nil {
-		return checkError(err, instance)
+	return nil
+}
+
+func applyServiceMesh(instance *servingv1alpha1.KnativeServing) error {
+	log.Info("Creating namespace for service mesh")
+	if err := createIngressNamespace(instance.GetNamespace()); err != nil {
+		return err
 	}
-	found := func(smmr *maistrav1.ServiceMeshMemberRoll) bool {
-		for _, member := range smmr.Status.ConfiguredMembers {
-			if member == "knative-serving" {
-				return true
+	log.Info(fmt.Sprintf("Successfully created namespace %s", ingressNamespace(instance.GetNamespace())))
+	log.Info("Installing serviceMeshControlPlane")
+	if err := installServiceMeshControlPlane(instance); err != nil {
+		return err
+	}
+	log.Info("Successfully installed serviceMeshControlPlane")
+	log.Info("Wait ServiceMeshControlPlane condition to be ready")
+	// wait for serviceMeshControlPlane condition to be ready before reconciling knative serving component
+	if err := waitForServiceMeshControlPlaneReady(instance.GetNamespace()); err != nil {
+		return err
+	}
+	log.Info("ServiceMeshControlPlane is ready")
+	log.Info("Installing ServiceMeshMemberRole")
+	if err := installServiceMeshMemberRole(instance.GetNamespace()); err != nil {
+		return err
+	}
+	log.Info(fmt.Sprintf("Successfully installed ServiceMeshMemberRole and configured %s namespace", instance.GetNamespace()))
+	log.Info(fmt.Sprintf("Wait ServiceMeshMemberRole to update %s namespace into configured members", instance.GetNamespace()))
+	if err := waitForServiceMeshMemberRoleReady(instance.GetNamespace()); err != nil {
+		return err
+	}
+	log.Info(fmt.Sprintf("Successfully configured %s namespace into configured members", instance.GetNamespace()))
+	return nil
+}
+
+// waitForServiceMeshControlPlaneReady checks whether serviceMeshControlPlane installs all required component
+func waitForServiceMeshControlPlaneReady(servingNamespace string) error {
```

Comment:
I think I might have concerns about polling within the reconcile loop of a KnativeServing CR. This seems wrong to me. Wouldn't it be better to return an error until the SMCP is ready? That way, we get the benefit of the controller's own rate-limited retry policy, and we don't back up the events queuing up to be reconciled.

Reply:
We considered that initially, but it felt wrong to me because the retry loop will be tight as opposed to controlled. We'd really need to return a `nil` early and setup a Watch against the SMCP/SMMR to trigger us again once that even changes.

I was okay with doing it the "wrong" way as it's not detrimental to any performance on our end. We only allow 1 KnativeServing resource anyway. This would be an issue if we allowed more though, I agree.

The plan was to implement the "proper" version in the operator you're currently building.

Reply:
My primary concern is if the user changes or deletes the KS CR before the polling is complete. If that reconcile loop is blocked, hijinks could ensue.